### ISSUE TITLE
fatal error during flambda pass

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#7291]
**Original bug ID:** 7291
**Reporter:** @c-cube
**Assigned to:** @mshinwell
**Status:** closed (set by @damiendoligez on 2016-10-26T14:26:19Z)
**Resolution:** fixed
**Priority:** normal
**Severity:** crash
**Platform:** all?
**OS:** archlinux 64 bits
**Version:** 4.03.0
**Target version:** 4.04.0 +dev / +beta1 / +beta2
**Fixed in version:** 4.04.0 +dev / +beta1 / +beta2
**Category:** middle end (typedtree to clambda)

## Bug description

A project of mine compiles fine on 4.03.0, but fails in 4.03.0+flambda with the following error:

```
>> Fatal error: When simplifying [Project_var], the closure ID subst/3626 in the approximation of the set of closures did not match the closure ID subst_partial_fun/3548 in the [Project_var] term.  Approximation: { descr=(closure:
subst/3626 from { descr=(set_of_closures:
subst/3626 invariant_params={ (param/3627 { }) } freshening={ vars_within_closure 
{ (id/3573 id/3625) }, closure_id { (subst/3574 subst/3626) } }) var=subst_set_of_closures/3632 symbol=<no symbol> }) var=subst/3636 symbol=<no symbol> }
 Var-within-closure being projected: subst_1967/3549

Fatal error: exception Misc.Fatal_error
```

## Steps to reproduce

$ opam sw 4.03.0+flambda
$ opam pin add -k git nunchaku https://github.com/nunchaku-inria/nunchaku.git#bug-flambda

^
it should fail during compilation of nunchaku itself. Sorry, the compilation is a bit long.

## Additional information

Opam files are attached (colored output enabled, use less -R).

## File attachments

- [opam-output.tar.gz](https://gist.githubusercontent.com/vicuna/42a874e81668adb320d22955b9180c11/raw/3b10f21353ea86f3b4a3446a861f514aed8d0840/opam-output.tar.gz)

**Comment by vicuna** [comment]
**Comment author:** @mshinwell

I am going to look at this.

**Comment by vicuna** [comment]
**Comment author:** @mshinwell

I have made progress on this today, and hope we can get a fix early next week.

**Comment by vicuna** [comment]
**Comment author:** @mshinwell

https://github.com/ocaml/ocaml/pull/731

**Comment by vicuna** [comment]
**Comment author:** @damiendoligez

This bug is fixed by #780, which supersedes #731.

https://github.com/ocaml/ocaml/pull/780
